CANVAS: a11y/Update (#21669)
mdn/content: files/en-us/learn/javascript/client-side_web_apis/drawing_graphics/index.md

This will create a canvas on the page with a size of 320 by 240 pixels.

> Inside the canvas tags, put fallback content. This will be shown if the user's browser doesn't support canvas, and information for screen readers.

Reviewer: ```suggestion
You should put some fallback content inside the canvas tags. This should describe the canvas content to users of browsers that don't support canvas, or users of screen readers.
```

> You should put some fallback content inside the canvas tags. This should describe the canvas content to users of browsers that don't support canvas, or users of screen readers.

Reviewer: ```suggestion
You should put some fallback content inside the `<canvas>` tags. This should describe the canvas content to users of browsers that don't support canvas, or users of screen readers.
```

```html
<canvas width="320" height="240">
  <p>Description of the canvas for those unable to view it.</p>
</canvas>
```

> The above message should provide useful fallback content to the canvas content. For example, if you were rendering a constantly updating graph of stock prices, the fallback content could be a static image of the latest stock graph, with `alt` text saying what the prices are in text or a list of links to individual stock pages.

Reviewer: ```suggestion
The fallback should provide useful alternative content to the canvas content. For example, if you are rendering a constantly updating graph of stock prices, the fallback content could be a static image of the latest stock graph, with `alt` text saying what the prices are in text or a list of links to individual stock pages.
```

> > **Note:** Canvas content is not accessible to screen readers. Include descriptive text as the value of the [`aria-label`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label) attribute directly on the canvas element itself or include fallback content placed within the opening and closing canvas tag. Canvas content is not part of the DOM, but nested fallback content is.

Reviewer: ```suggestion
> **Note:** Canvas content is not accessible to screen readers. Include descriptive text as the value of the [`aria-label`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label) attribute directly on the canvas element itself or include fallback content placed within the opening and closing `<canvas>` tags. Canvas content is not part of the DOM, but nested fallback content is.
```

There are also a number of properties to help control text rendering such as {{domxref("CanvasRenderingContext2D.font", "font")}}, which lets you specify font family, size, etc. It takes as its value the same syntax as the CSS {{cssxref("font")}} property.

> 
> Canvas content is not accessible to screen readers. Text painted to the canvas is not available to the DOM, but must be made available to be accessible. In this example, we include the text as the value for `aria-label`.

Reviewer: ```suggestion
Canvas content is not accessible to screen readers. Text painted to the canvas is not available to the DOM, but must be made available to be accessible. In this example, we include the text as the value for `aria-label`.
```

> 

Reviewer: Seems duplicated from the note above.

Author: It is slightly redundant, but as alternative text is vital in the case of text in canvas, it is ok to be slightly redundant. the case here is different: all the other examples use inner text. this one is using aria-label because it is text.

ctx.fillStyle = "red";
ctx.font = "48px georgia";
ctx.fillText("Canvas text", 50, 150);

canvas.setAttribute("aria-label", "Canvas text");
```

Here we draw two lines of text, one outline and the other stroke. The final example should look like so:

{{EmbedGHLiveSample("learning-area/javascript/apis/drawing-graphics/getting-started/4_canvas_text/index.html", '100%', 180)}}

> > **Note:** The finished code is available on GitHub as [4_canvas_text](https://github.com/mdn/learning-area/tree/main/javascript/apis/drawing-graphics/getting-started/4_canvas_text). 

Reviewer: ```suggestion
> **Note:** The finished code is available on GitHub as [4_canvas_text](https://github.com/mdn/learning-area/tree/main/javascript/apis/drawing-graphics/getting-started/4_canvas_text).
```

Now let's create our own simple animation — we'll get a character from a certain rather awesome retro computer game to walk across the screen.

1. Make another fresh copy of our canvas template ([1_canvas_template](https://github.com/mdn/learning-area/tree/main/javascript/apis/drawing-graphics/getting-started/1_canvas_template)) and open it in your code editor.
   

> 2. Update the inner HTML to reflect the image:
>    ```html
>     <canvas class="myCanvas">
>       <p>A man walking.</p>
>     </canvas>
>    ```

Reviewer: ````suggestion
2. Update the inner HTML to reflect the image:

   ```html
    <canvas class="myCanvas">
      <p>A man walking.</p>
    </canvas>
   ```

````

3. At the bottom of the JavaScript, add the following line to once again make the coordinate origin sit in the middle of the canvas:

   ```js
   ctx.translate(width / 2, height / 2);
   ```

> 4. Now let's create a new {{domxref("HTMLImageElement")}} object, set its {{htmlattrxref("src", "img")}} to the image we want to load and add an `onload` event handler that will cause the `draw()` function to fire when the image is loaded:

Reviewer: ```suggestion
4. Now let's create a new {{domxref("HTMLImageElement")}} object, set its {{htmlattrxref("src", "img")}} to the image we want to load, and add an `onload` event handler that will cause the `draw()` function to fire when the image is loaded:
```

(There are three actions here, which is why the comma is necessary)

   If our character hasn't yet walked off the edge of the screen, we increment `posX` by 2. This will make him move a little bit to the right the next time we draw him.

11. Finally, we need to make the animation loop by calling {{domxref("window.requestAnimationFrame", "requestAnimationFrame()")}} at the bottom of the `draw()` function:

>    ```js

Reviewer: ```suggestion
    ```js
```
